rtc_auth_enclave/src/lib.rs

```diff
@@ -3,9 +3,116 @@
 #![deny(unsafe_op_in_unsafe_fn)]
 #![deny(clippy::mem_forget)]
 
+mod jwt;
+mod token_store;
+
 #[cfg(not(target_env = "sgx"))]
 extern crate sgx_tstd as std;
 
+use core::slice;
+use std::ptr;
+use std::string::{String, ToString};
+
+use rtc_tenclave::crypto::{RtcCrypto, SodaBoxCrypto as Crypto};
 pub use rtc_tenclave::dh::*;
 #[allow(unused_imports)] // for ECALL linking
 use rtc_tenclave::enclave::enclave_create_report;
+use rtc_types::{EcallResult, EncryptedMessage, ExecReqMetadata, ExecTokenError, IssueTokenResult};
+use secrecy::{ExposeSecret, Secret};
+use serde::{Deserialize, Serialize};
+use uuid::Uuid;
+
+#[derive(Serialize, Deserialize)]
+pub struct ExecReqData {
+    dataset_uuid: [u8; 16],
+    dataset_access_key: [u8; 24],
+    exec_module_hash: [u8; 32],
+    number_of_uses: u32,
+}
+
+pub type Nonce = [u8; 24];
+
+/// Issue an execution token using the parameters defined in the payload.
+///
+/// # Safety
+/// This function expects
+///   1. `payload_ptr` to be valid for a slice of len `payload_len`
+///   2. `metadata_ptr` should be a valid pointer.
+///   3. An allocated buffer of size `out_token_len` that starts at `out_token_ptr`
+/// The edge code from sgx must uphold the above.
+#[no_mangle]
+pub unsafe extern "C" fn issue_execution_token(
+    payload_ptr: *const u8,
+    payload_len: usize,
+    metadata_ptr: *const ExecReqMetadata,
+    out_token_ptr: *mut u8,
+    out_token_len: usize,
```

Comment:
So with this the caller would do something like `out_buffer[out_token_len..]` to get the actual data. That make sense, since I ran into issues with the authenticated encryption and padded 0s.

Reply:
Yeah; it's a pattern I've seen a few times in SGX discussions for bounded variable-sized returns.